### ISSUE TITLE
packages from repo.powerdns.com for trusty tahr don't serve AXFR for a signed zone on AWS

### ISSUE THREAD

**Issue by ahupowerdns** [PowerDNS/pdns#3321]
So I realize this is awfully specific, and I can't reproduce except via the packages from repo.powerdns.com. To reproduce, get an AWS 'trusty' image, install pdns from repo.powerdns.com with gsqlite3 backend. Create a zone with pdnsutil (rock on), do pdnsutil secure-zone, attempt to AXFR it. It will lock up. 

If you compile on the AWS host itself there is no problem. Idem on my 'real' trusty tahr. Haven't yet tried the repo.powerdns.com packages locally.

The problem happens after:
```
log.13134-access("/etc/powerdns/powerdns.sqlite3-wal", F_OK) = -1 ENOENT (No such file or directory)
log.13134-fstat(23, {st_mode=S_IFREG|0664, st_size=1321984, ...}) = 0
log.13134-fcntl(23, F_SETLK, {type=F_UNLCK, whence=SEEK_SET, start=0, len=0}) = 0
log.13134:futex(0x7f149e129438, FUTEX_LOCK_PI_PRIVATE, 1) = -1 ESRCH (No such process)
log.13134-pause( <unfinished ...>
log.13134-+++ killed by SIGINT +++
```

On "working" pdns_server, that 'LOCK_PI_PRIVATE' doesn't happen, nor the pause. This is the thread that serves the AXFR. The issue is somewhere after   dk.getFromMeta(q->qdomain, "PUBLISH_CDS", publishCDS);

**Comment by ahupowerdns** [comment]
potentially related, the "ubuntu trusty tahr" on AWS is different from mine..

**Comment by ahupowerdns** [comment]
https://lkml.org/lkml/2014/5/14/52 has some words on it - this is glibc giving up on a lock. Unsure why this happens. sqlite does use recursive mutexes which are implemented as 'FUTEX_LOCK_PI'.

**Comment by pieterlexis** [comment]
Confirmed this is also an issue on non-AWS trusty installations. This issue is not only on AXFR it is also on any query with the DO-bit set.

**Comment by ahupowerdns** [comment]
is our trusty build environment up to date? it appears to be using kernel features that aren't used on trusty-on-trusty compilations.

**Comment by ahupowerdns** [comment]
THE PAIN - something locks weirdly within openssl
```
(gdb) bt
#0  0x00007f014e8c1c40 in __pause_nocancel () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007f014e8b7ce6 in __pthread_mutex_lock_full (mutex=0x7f0151202408)
    at ../nptl/pthread_mutex_lock.c:354
#2  0x00007f014f586c71 in EC_KEY_get_key_method_data () from /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#3  0x00007f014f5a572d in ?? () from /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#4  0x00007f014f5a65ff in ECDSA_do_sign_ex () from /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#5  0x00007f0150d677d2 in OpenSSLECDSADNSCryptoKeyEngine::sign(std::string const&) const ()
#6  0x00007f0150c53770 in fillOutRRSIG(DNSSECPrivateKey&, DNSName const&, RRSIGRecordContent&, std::vector<std::shared_ptr<DNSRecordContent>, std::allocator<std::shared_ptr<DNSRecordContent> > >&) ()
#7  0x00007f0150c54614 in getRRSIGsForRRSET(DNSSECKeeper&, DNSName const&, DNSName, unsigned short, unsigned int, std::vector<std::shared_ptr<DNSRecordContent>, std::allocator<std::shared_ptr<DNSRecordContent> > >&, std::vector<RRSIGRecordContent, std::allocator<RRSIGRecordContent> >&) ()
#8  0x00007f0150c54b58 in addSignature(DNSSECKeeper&, UeberBackend&, DNSName const&, DNSName, DNSName const&, unsigned short, unsigned int, DNSResourceRecord::Place, std::vector<std::shared_ptr<DNSRecordContent>, std::allocator<std::shared_ptr<DNSRecordContent> > >&, std::vector<DNSResourceRecord, std::allocator<DNSResourceRecord> >&, unsigned int) ()
#9  0x00007f0150c55eb1 in addRRSigs(DNSSECKeeper&, UeberBackend&, std::set<DNSName, std::less<DNSName>, std::allocator<DNSName> > const&, std::vector<DNSResourceRecord, std::allocator<DNSResourceRecord> >&) ()
#10 0x00007f0150cb0a61 in PacketHandler::questionOrRecurse(DNSPacket*, bool*) ()
#11 0x00007f0150cb20ad in PacketHandler::question(DNSPacket*) ()
#12 0x00007f0150bf2d8e in MultiThreadDistributor<DNSPacket, DNSPacket, PacketHandler>::makeThread(void*) ()
#13 0x00007f014e8ba182 in start_thread (arg=0x7f01496e9700) at pthread_create.c:312
#14 0x00007f014e5e747d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

**Comment by ahupowerdns** [comment]
fixed in 305c6ab

**Comment by ahupowerdns** [comment]
for posterity, this was caused by a mistake we made which meant we only initialized half the locks in openssl. Valgrind would normally pick up such things, but not this time since openssl has.. its own memory allocator. Yes. Eventually with sufficient shuffling we did manage to make valgrind see the problem.
